src/Chapter3.hs

```diff
@@ -1006,6 +1142,189 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+-- Couldn't make it simpler :(
+data KnightAction
+  = KnightAttack
+  | Drink
+  | Cast
+  deriving (Show)
+
+data MonsterAction
+  = MonsterAttack
+  | Run
+  deriving (Show)
+
+data Knight = Knight
+  { knightHealth :: Int
+  , knightAttack :: Int
+  , knightDefense :: Int
+  , knightActions :: [KnightAction]
+  , baseKnightHealth :: Int
+  , baseKnightHealthIncrease :: Int
+  , baseKnightDefense :: Int
+  , baseKnightDefenseIncrease :: Int
+  } deriving (Show)
+
+data Monster = Monster
+  { monsterHealth :: Int
+  , monsterAttack :: Int
+  , monsterActions :: [MonsterAction]
+  } deriving (Show)
+
+data ActionType
+  = GenericAttack
+  | SpecialAction
+  deriving (Show)
+
+getFirstKnightAction :: Knight -> KnightAction
+getFirstKnightAction Knight { knightActions = (x:_) } = x
+getFirstKnightAction _ = error "Forgot to 'cycle' knight actions"
```

Comment:
It is not a nice practice to use `error` function in production code. And most of the time, you can omit that by using better types 🙂 
To avoid `error` function in here, for example, one option is to use `NonEmpty` data type instead of list. `NonEmpty` is almost like list but it guaranteed to have at least 1 element 🙂

Reply:
Seemed like a hack, but I'm not experienced with such type system, so tried to do what I know. This task was quite hard as it is :-)

Reply:
Yeah, I know, you did great! 👍🏼